src/models/geograph.py

```diff
@@ -443,27 +450,143 @@ def merge_nodes(
             and self.graph.has_node(final_index)
         ):
             raise ValueError(
-                "`final_index` must not be an existing node not in `node_list`."
+                "`final_index` cannot be an existing node that is not in `node_list`."
             )
         if final_index is None:
-            final_index = self.graph.number_of_nodes
+            final_index = self.df.last_valid_index()
 
-        adjacency_list = set()
+        # Build set of all neighbours of nodes in node_list, excluding the
+        # nodes in node_list.
+        adjacency_set = set()
         for node in node_list:
-            adjacency_list.update(self.graph.neighbors(node))
-        polygon = unary_union(
-            [self.graph.nodes[node]["geometry"] for node in node_list]
-        )
-
+            adjacency_set.update(list(self.graph.neighbors(node)))
+        adjacency_set -= adjacency_set.intersection(node_list)
+        # Build union polygon.
+        polygon = self.df.geometry.iloc[node_list].unary_union
+        # Remove nodes from graph and rows from df
         self.graph.remove_nodes_from(node_list)
+        self.df = self.df.drop(node_list)
+        # Add final node to graph and df
         self.graph.add_node(
             final_index,
             rep_point=polygon.representative_point(),
             area=polygon.area,
             perimeter=polygon.length,
-            geometry=polygon,
             class_label=class_label,
+            bounds=polygon.bounds,
+        )
+        data = {"class_label": class_label, "geometry": polygon}
+        missing_cols = {key: None for key in set(self.df.columns) - set(data.keys())}
+        self.df.loc[final_index] = gpd.GeoSeries(
+            {**data, **missing_cols}, crs=self.df.crs
         )
+        self.df = self.df.sort_index()
+        # Add edges from final node to all neighbours of the original
         self.graph.add_edges_from(
-            zip_longest([final_index], adjacency_list, fillvalue=final_index)
+            zip_longest([final_index], adjacency_set, fillvalue=final_index)
+        )
+
+    def add_habitat(
+        self,
+        name: str,
+        valid_classes: set,
+        max_travel_distance: float = 0.0,
+    ) -> None:
+        """
```

Comment:
Super nice function! 🎉  Thanks a lot for implementing it, will be great for a lot of other applications.

This fairly nitpicking but we could consider also accepting `list` besides `set` as the type for the `valid_classes`. My thinking here is that many users (including myself) aren't too familiar with set, so it might add a little confusion for them. It also shouldn't be difficult to do because both because you can use the `set` operator on a set and a list, like `set(set([1,0,1]))`. So just adding the type in the docstring and the line below would do it (as far as I understand it):
```
valid_classes = set(valid_classes)
```

Reply:
One more thing: would it also be possible to save the inputs for the habitat? It might be handy to have both the max_travel_distance and valid_classes somewhere. (I am thinking to create a visualisation later). Maybe the habitat property could be a list of dict of dicts 
```
{name:{graph:nx.graph,max_travel_distance:int,valid_classes:set},...}
```
It's easier to change this now than later ... but maybe you have a better idea of how to store this info?

Reply:
Done!